### ISSUE TITLE
Machine.AddHostAlias fails when run in DinD

### ISSUE THREAD

**Issue by stiller-leser** [kubernetes/minikube#8273]
**Steps to reproduce the issue:** 

1. Start https://github.com/bsycorp/kind with `MINIKUBE_VERSION=latest` and `KUBERNETES_VERSION=v1.18.2`
2. This starts minikube with the following start command: `minikube start '--vm-driver=none' --kubernetes-version v1.18.2 --bootstrapper kubeadm --apiserver-ips 172.30.99.1,127.0.0.1 --apiserver-name minikube '--extra-config=apiserver.advertise-address=172.30.99.1' '--extra-config=kubeadm.ignore-preflight-errors=FileContent--proc-sys-net-bridge-bridge-nf-call-iptables,Service-Docker'`

**Full output of failed command:** 

```
* minikube v1.10.1 on Alpine 3.11.3 (docker/amd64)
  - MINIKUBE_VERSION=latest
* Using the none driver based on user configuration
* Starting control plane node minikube in cluster minikube
* Running on localhost (CPUs=4, Memory=16024MB, Disk=64474MB) ...
* OS release is Alpine Linux v3.11
* Preparing Kubernetes v1.18.2 on Docker 19.03.5 ...
E0526 09:27:26.658421     189 start.go:97] Unable to add host alias: hosts update: /bin/bash -c "{ grep -v '\thost.minikube.internal$' /etc/hosts; echo "127.0.0.1	host.minikube.internal"; } > /tmp/h.$$; sudo cp /tmp/h.$$ /etc/hosts": exit status 1
stdout:

stderr:
cp: can't create '/etc/hosts': File exists
  - apiserver.advertise-address=172.30.99.1
  - kubeadm.ignore-preflight-errors=FileContent--proc-sys-net-bridge-bridge-nf-call-iptables,Service-Docker
    > kubectl.sha256: 65 B / 65 B [--------------------------] 100.00% ? p/s 0s    > kubelet.sha256: 65 B / 65 B [--------------------------] 100.00% ? p/s 0s    > kubeadm.sha256: 65 B / 65 B [--------------------------] 100.00% ? p/s 0s    > kubeadm: 7.25 MiB / 37.97 MiB [---->_____________________] 19.09% ? p/s ?    > kubectl: 3.45 MiB / 41.99 MiB [-->________________________] 8.22    > kubelet: 25.80 MiB / 108.03 MiB [----->__________________] 23.88% ? p/s ?% ? p/s ?    > kubeadm: 13.94 MiB / 37.97 MiB [--------->_______________] 36.71% ? p/s ?    > kubectl: 19.47 MiB / 41.99 MiB [----------->_____________] 46.36% ? p/s ?    > kubelet: 44.14 MiB / 108.03 MiB [--------->______________] 40.86% ? p/s ?    > kubeadm: 37.97 MiB / 37.97 MiB [--------------] 100.00% 108.94 MiB p/s 0s    > kubectl: 41.99 MiB / 41.99 MiB [--------------] 100.00% 117.01 MiB p/s 0s    > kubelet: 46.69 MiB / 108.03 MiB [---------->_____________] 43.22% ? p/s ?    > kubelet: 81.42 MiB / 108.03 MiB [-------->__] 75.37% 92.71 MiB p/s ETA 0s    > kubelet: 108.03 MiB / 108.03 MiB [------------] 100.00% 151.14 MiB p/s 1s* 
X Failed to update cluster: updating node: host alias: hosts update: /bin/bash -c "{ grep -v '\tcontrol-plane.minikube.internal$' /etc/hosts; echo "172.17.0.2	control-plane.minikube.internal"; } > /tmp/h.$$; sudo cp /tmp/h.$$ /etc/hosts": exit status 1
stdout:

stderr:
cp: can't create '/etc/hosts': File exists

* 
* minikube is exiting due to an error. If the above message is not useful, open an issue:
  - https://github.com/kubernetes/minikube/issues/new/choose
```
In [Node.Start](https://github.com/kubernetes/minikube/blob/07c7b7ca9771bef3790ec6d1b20546eee1b4b849/pkg/minikube/node/start.go#L93) it fails intentionally gracefully. However, in [Machine.AddHostAlias](https://github.com/kubernetes/minikube/blob/07c7b7ca9771bef3790ec6d1b20546eee1b4b849/pkg/minikube/machine/start.go#L275) it crashes directly.

I was wondering if that can also fail gracefully with an hint, requesting the user to add the entry manually before running minikube?

Best regards,
stiller-leser

**Comment by tstromberg** [comment]
Interesting. I haven't looked at https://github.com/bsycorp/kind in years, but it's worth noting that minikube now has the functionality built-in:

`minikube start --driver=docker`

With the recent changes within minikube, we do have a hard requirement on being able to edit `/etc/hosts`. We do this so that the cluster continues to work if the underlying IP changes.  Are you trying to run minikube within a Kubernetes cluster? Or is there something else that is going on that prevents /etc/hosts from being edited within the container?

If it's within a cluster, you can get by this by adding a control-plane host alias: https://kubernetes.io/docs/concepts/services-networking/add-entries-to-pod-etc-hosts-with-host-aliases/

**Comment by stiller-leser** [comment]
RedHat is using as the default way to test ansible operators via Molecule in the OperatorSDK. I suppose I could run a Minikube cluster manually before executing the tests, but the current way allows for quite a nice experience.

**Comment by tstromberg** [comment]
@stiller-leser - are you running minikube within a Kubernetes cluster with this test?

**Comment by afbjorklund** [comment]
As far as I know, this situation is true in _any_ type of docker container:

```console
$ docker run -it alpine
/ # cp /etc/hosts hosts
/ # cp hosts /etc/hosts
cp: can't create '/etc/hosts': File exists
```

It has some magic mounting of certain system files from the host:

```
/etc/resolv.conf
/etc/hostname
/etc/hosts
```


I guess we could use something like `grep /etc/hosts /proc/mounts` and pretend that we support the "none" driver inside a container (we don't, since kubeadm doesn't). But like LXC, "while stock lasts" ?

Some of this was discussed in #7957

**Comment by stiller-leser** [comment]
No, molecule starts the kind container (which in turn uses minikube's components to start a cluster) on my desktop . The tests are then executed against the cluster running in the docker container on my desktop. The way I read the kind code is that minikube is only used to download the k8s components and configure the certs, etc... Kind itself uses a bunch of custom scripts.


The issue I originally describes prevents the download and cert creation to happen at all, breaking kind's "installation" logic. I agree that this is not directly a minikube issue. I just saw the two ambivalent error handlings and was wondering if they could both fail gratefully. As far as I understood the kind code, it would stop working otherwise in the future. 




Gesendet von Outlook Mobile







On Tue, May 26, 2020 at 10:30 PM +0200, "Thomas Strömberg" <notifications@github.com> wrote:















@stiller-leser - are you running minikube within a Kubernetes cluster with this test?

**Comment by afbjorklund** [comment]
Guess we could get creative, and try to avoid the "cp" command (and friends, like `sed -i`)

```
/ # cp /etc/hosts hosts
/ # ( cat hosts; echo 127.0.1.1 kilroy ) | tee /etc/hosts
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
172.17.0.3	d5cb6d85d922
127.0.1.1 kilroy
```

That is, to edit the file in place (somehow, tee is popular) rather than trying to replace it ?

**Comment by medyagh** [comment]
I don't think we have ever tested minikube run in Docker in Docker using Docker Driver. thats 3 Layers of docker in docker

**Comment by stiller-leser** [comment]
I understand that this is a strange concept and not something that seem to make any sense in the first place. However, it is extremly usefull during the development of an operator. It allows for fast and easy local testing. Same for CI use cases. I agree that nobody should do something like this in production, though ;)

**Comment by medyagh** [comment]
it is technically possible, only is you would need to create the first container as a privileged

**Comment by afbjorklund** [comment]
> it is technically possible, only is you would need to create the first container as a privileged

Not sure it makes any difference ?

```console
$ docker run -it alpine
/ # cp /etc/hosts hosts
/ # cp hosts /etc/hosts
cp: can't create '/etc/hosts': File exists
$ docker run -it --privileged alpine
/ # cp /etc/hosts hosts
/ # cp hosts /etc/hosts
cp: can't create '/etc/hosts': File exists
```

For minikube, we need to replace this code:

```go
        script := fmt.Sprintf(`{ grep -v '\t%s$' /etc/hosts; echo "%s"; } > /tmp/h.$$; sudo cp /tmp/h.$$ /etc/hosts`, name, record)
        if _, err := c.RunCmd(exec.Command("/bin/bash", "-c", script)); err != nil {
                return errors.Wrap(err, "hosts update")
        }
```

With something that modifies the file in-place

**Comment by polarathene** [comment]
> ```console
> cp: can't create '/etc/hosts': File exists
> ```

Just to clarify regarding discussions about `cp`, this is specific to the BusyBox implementation of `cp` which will change the inode when the destination differs from the source mount/filesystem.

Compared to Coreutils `cp`, if you try with Fedora you'll be prompted to overwrite (_unless using `-f` / `--force`_), or Ubuntu (_no overwrite prompt, note that this distro plans to adopt `uutils` in future instead_). So relying on this command is a bit inconsistent, plus can involve an intermediate copy, a better approach is shown below with `sponge`

I'm not sure if my comment adds value much here, since it's already been discussed that the solution is to do an in in-place overwrite (_which confusingly is not what `sed --in-place` does_ 🙄).

---

> Guess we could get creative, and try to avoid the "cp" command (and friends, like `sed -i`)
> ```
> cp /etc/hosts hosts
> ( cat hosts; echo 127.0.1.1 kilroy ) | tee /etc/hosts
> ```
> That is, to edit the file in place (somehow, tee is popular) rather than trying to replace it?

As mentioned `sed -i` would not work. If you did use `sed` to avoid a copy for whatever modification you want to make to `/etc/hosts`, do not pair it with `tee` however, as read and writing from the same file will occur at the same time which is racey.

A common solution for such is `sponge` from the `moreutils` package, which will avoid writing until it's finished reading from stdin (_such as a `sed` call streaming output_).

You can use it like this:

```bash
export WITH_HOSTNAME='hello.world.test'
sed -E "s|($(cat /proc/sys/kernel/hostname))|${WITH_HOSTNAME} \1|" /etc/hosts | sponge /etc/hosts
```

**Full example for reference:**

```console
$ docker run --rm -it --hostname example.test fedora:42
$ dnf instally -yq hostname

# `hostname` and `hostname --fqdn` have the same output at this point, but:
# - The `hostname` command should output the value of `/proc/sys/kernel/hostname` (aka `sysctl kernel.hostname`)
# - The `--fqdn` option relies on resolution via a glibc lookup that uses `/etc/nsswitch.conf` to query the value of `/proc/sys/kernel/hostname`
$ hostname
example.test
$ hostname --fqdn
example.test

# Here we're prepending our preferred hostname to the entry in `/etc/hosts`
# that Docker already added our containers configured hostname with:
$ export WITH_HOSTNAME='hello.world.test'
$ sed -E "s|($(cat /proc/sys/kernel/hostname))|${WITH_HOSTNAME} \1|" /etc/hosts | sponge /etc/hosts

# Now any software that uses libc with NSS to resolve the container hostname FQDN
# will get the preferred FQDN we've added to `/etc/hosts` (matches output of `hostname --fqdn`).
# NOTE:
# - This cannot change the actual hostname assigned to the UTS namespace of the kernel for this container (/proc/sys/kernel/hostname).
# - Nor does it update the `HOSTNAME` ENV.

# Some methods of this command don't interact with NSS, so their output remains the same:
$ hostname
example.test
$ hostname --short
example

# While others now reflect the preferred FQDN:
$ hostname --fqdn
hello.world.test
$ hostname --domain
world.test
```

AFAIK Kubernetes lacks the ability to configure a containers hostname (_like shown with Docker via `--hostname`_).

The approach above is what I use to ensure an FQDN when software relies on making libc calls to acquire the hostname/FQDN for the container from `/etc/hosts` / DNS (_managed via `/etc/nsswitch.conf`_), one example of this was Fetchmail configuring it's HELO (_without `--hostname` to configure a FQDN hostname for the container, some MTAs would reject the connection_).

---

**NOTE:** `moreutils` / `sponge` is not exactly lightweight for containers if they don't already have deps like Perl installed (_Alpine as a base image at 8.1 MB grows to 49.8 MB_). There are some rewrites/alternatives written in Go ([`hymkor/sponge`](https://github.com/hymkor/sponge), [`shmokmt/sponge-go`](https://github.com/shmokmt/sponge-go)) and Rust ([`hymkor/sponge-rs`](https://github.com/hymkor/sponge-rs), [`jridgewell/rw`](https://github.com/jridgewell/rw)), which could be used as smaller static binaries.
